packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,431 @@
+  /// leaves out support for input method editors, and doesn't support soft
+  /// keyboards in general. For text input, consider [TextField],
+  /// [EditableText], or [CupertinoTextField] instead, which do support these
+  /// things.
+  final FocusOnKeyCallback onKey;
+
+  /// Handler called when the focus changes.
+  ///
+  /// Called with true if this node gains focus, and false if it loses
+  /// focus.
+  final ValueChanged<bool> onFocusChange;
+
+  /// True if this widget will be selected as the initial focus when no other
+  /// node in its scope is currently focused.
+  final bool autofocus;
+
+  /// An optional focus node to use as the focus node for this [Focus] widget.
+  ///
+  /// If one is not supplied, then one will be allocated and owned by this
+  /// widget.
+  ///
+  /// Supplying a focus node is sometimes useful if an ancestor to this widget
+  /// wants to control when this widget has the focus. The owner will be
+  /// responsible for calling [FocusNode.dispose] on the focus node when it is
+  /// done with it, but this [Focus] widget will attach/detach and reparent the
+  /// node when needed.
+  final FocusNode focusNode;
 
-  /// A list of the [FocusScopeNode]s for each [FocusScope] ancestor of
-  /// the given [BuildContext]. The first element of the list is the
-  /// nearest ancestor's [FocusScopeNode].
+  /// Returns the [focusNode] of the [Focus] that most tightly encloses the given
+  /// [BuildContext].
   ///
-  /// The returned list does not include the [FocusManager]'s `rootScope`.
-  /// Only the [FocusScopeNode]s that belong to [FocusScope] widgets are
-  /// returned.
+  /// If this node doesn't have a [Focus] widget ancestor, then the
+  /// [FocusManager.rootScope] is returned.
   ///
   /// The [context] argument must not be null.
-  static List<FocusScopeNode> ancestorsOf(BuildContext context) {
+  static FocusNode of(BuildContext context) {
     assert(context != null);
-    final List<FocusScopeNode> ancestors = <FocusScopeNode>[];
-    while (true) {
-      context = context.ancestorInheritedElementForWidgetOfExactType(_FocusScopeMarker);
-      if (context == null)
-        return ancestors;
-      final _FocusScopeMarker scope = context.widget;
-      ancestors.add(scope.node);
-      context.visitAncestorElements((Element parent) {
-        context = parent;
-        return false;
-      });
-    }
+    final _FocusMarker marker = context.inheritFromWidgetOfExactType(_FocusMarker);
+    return marker?.notifier ?? context.owner.focusManager.rootScope;
   }
 
+  /// Returns true if the nearest enclosing [Focus] widget's node is focused.
+  ///
+  /// A convenience method to allow build methods to write:
+  /// `Focus.hasFocus(context)` to get whether or not the nearest [Focus] above
+  /// them in the widget hierarchy currently has the keyboard focus.
+  static bool hasFocus(BuildContext context) => Focus.of(context).hasFocus;
```

Comment:
This doesn't read very well in usage, IMHO.

`if (Focus.hasFocus(context))` reads as "If the focus has the focus, context", which doesn't make much sense.

That's why I liked 'at`:

`if (Focus.at(context))` reads as "If the focus is at context", which is exactly what you are asking.

Reply:
My 5 cents, maybe `if (Focus.isAt(context))`? Reads well

Reply:
Thanks, I do like `isAt` better.  Changed to that.